orttraining/orttraining/python/training/ortmodule/_training_manager.py

```diff
@@ -345,3 +345,22 @@ def _reinitialize_graph_builder(self, input_info):
             self._initialize_graph_builder(training=True)
             return True
         return False
+
+    def __getstate__(self):
+        # Attempt to serialize training manager
+
+        state = super(TrainingManager, self).__getstate__()
+
+        # Only top level classes are pickleable. So, _ORTModuleFunction is
+        # not pickleable. So, let's not pickle it, and redefine it when
+        # loading the state.
+        state['_forward_class'] = None
+        return state
+
+    def __setstate__(self, state):
+        # Attempt to deserialize training manager
+
+        super(TrainingManager, self).__setstate__(state)
+
+        # Redefine self._forward_class
+        self._forward_class = self._create_autofunction_class()
```

Comment:
I'm wondering if we should change the design. self._create_autofunction_class is wrong in a way it creates a class having self as a hidden context. Then the static method ``def forward(ctx, inputs)`` has two contexts: ctx and self. A better design would be to add self to the context ctx.

```
def ort_forward(ctx, *inputs):
    """
    Implements forward function.
    See :epkg:`autograd functions`.
    """
    cls = ctx._forward_cls  # it is the class returned by _create_autofunction_class
    # ...
    return ...

def _create_autofunction_class(self):

        kwargs = {
            '__doc__': doc,
            '__module__': __name__,
            '_run_options': self.run_options,
            '_sess': sess,
            '_sess_eval': sess_eval,
            '_training_agent': training_agent,
            '_cache': OrtValueCache(),
            '_update_cache': False,
            '_states': [],
            '_logger': logger,
            '_input_names': self.input_names,
            '_debug': debug,
            '_grad_input_names': grad_input_names,
            '_output_names': self.output_names,
            '_bw_fetches_names': bw_fetches_names,
            '_fw_outputs_device_info': fw_outputs_device_info,
            '_bw_outputs_device_info': bw_outputs_device_info,
            '_fw_no_grad_output_device_info': fw_no_grad_output_device_info,
            '_weights_to_train': list(sorted(
                self.weights_to_train)),
            'forward': staticmethod(ort_forward),
            'backward': staticmethod(ort_backward),
            '_graph_info': graph_info}

        if keep_models:
            kwargs.update(dict(
                _trained_onnx=onnx.load(BytesIO(train_onnx_model_serialized)),
                _optimized_pre_grad_model=onnx.load(
                    BytesIO(optimized_pre_grad_model)),
                _graph_builder=builder,
                _factory=self))

        newclass = type(self.class_name, (Function,), kwargs)
        return newclass
```

That would avoid patching forward or backward methods.

Reply:
We can do that. Another idea I was thinking earlier but didn't implement was to use `functools.partial` to bind an argument (`training_manager` in this case) into the function arguments, so that the methods have access to the `training_manager`.

My concern with this idea is that it adds complexity to the code that is not very nice for code readers. If we think this is the way forward, I can make the changes.